src/core/unif.ml

```diff
@@ -61,19 +60,24 @@ let nl_distinct_vars : term array -> tvar array option =
   try Some (Array.map replace_nl_var (Array.map to_var ts))
   with Not_a_var -> None
 
-(** [instantiate m ts u] check whether [m] can be instantiated for solving the
-    unification problem “m[ts] = u”. The returned boolean tells whether [m]
-    was instantiated or not. *)
+(** [instantiation m ts u] checks whether, in a problem [m[ts]=u], [m] can be
+   instantiated and returns the corresponding instantiation. It does not check
+   whether the instantiation is closed though. *)
+let instantiation : meta -> term array -> term ->
+                    (term, term) Bindlib.mbinder Bindlib.box option
+  = fun m ts u ->
+  if (!can_instantiate || internal m) && not (occurs m u) then
+    match nl_distinct_vars ts with
+    | None -> None
+    | Some vs -> Some (Bindlib.bind_mvar vs (lift u))
+  else None
+
+(** [instantiate m ts u] check whether, in a problem [m[ts]=u], [m] can be
+   instantiated and, if so, instantiate it. *)
 let instantiate : meta -> term array -> term -> bool = fun m ts u ->
-  (!can_instantiate || internal m)
-  && not (occurs m u)
-  && match nl_distinct_vars ts with
-     | None -> false
-     | Some vs ->
-        let bu = Bindlib.bind_mvar vs (lift u) in
-        Bindlib.is_closed bu (* To make sure that there is no non-linear
-                                variable of [vs] occurring in [u]. *)
-        && (set_meta m (Bindlib.unbox bu); true)
+  match instantiation m ts u with
+  | Some bu when Bindlib.is_closed bu -> set_meta m (Bindlib.unbox bu); true
+  | _ -> false
```

Comment:
I like the idea of unboxing as late as possible, and only when closed. In that case this is actually a no-op.